unicorn/src/unicorn/zookeeper.cpp

```diff
@@ -42,53 +34,570 @@
+                } catch(const std::system_error& e) {
+                    COCAINE_LOG_ERROR(parent->parent.log, "can not delete lock on {} - {}, reconnecting as a last resort",
+                                      lock_path, error::to_string(e));
+                    parent->parent.zk.reconnect();
+                }
+                parent = nullptr;
+            }
+        }
+    };
+
+private:
+    zookeeper_t& parent;
+    path_t folder;
+    path_t path;
+    path_t created_sequence_node;
+    value_t value;
+    uint64_t depth;
+
+public:
+    lock_t(callback::lock wrapped, zookeeper_t& parent, path_t folder) :
+        safe(std::make_shared<lock_scope_t>(), std::move(wrapped)),
+        parent(parent),
+        folder(std::move(folder)),
+        path(this->folder + "/lock"),
+        value(time(nullptr)),
+        depth(0)
+    {}
+
+    auto run() -> void {
+        //TODO: What can we do with this ugly hack with dynamic_cast?
+        std::dynamic_pointer_cast<lock_scope_t>(scope())->parent = std::dynamic_pointer_cast<lock_t>(shared_from_this());
+        COCAINE_LOG_DEBUG(parent.log, "starting lock operation on path {}, folder {}", path, folder);
+        parent.zk.create(path, serialize(value), true, true, shared_from_this());
+    }
+
+private:
+    auto on_reply(create_reply_t reply) -> void override {
+        if(reply.rc == ZOK) {
+            if(depth == 0) {
+                return scope()->closed.apply([&](bool& closed){
+                    created_sequence_node = get_node_name(reply.created_path);
+                    if(!closed) {
+                        parent.zk.childs(folder, shared_from_this());
+                    } else {
+                        parent.zk.del(created_sequence_node, shared_from_this());
+                    }
+                });
+            } else if(depth == 1) {
+                depth--;
+                parent.zk.create(path, serialize(value), true, true, shared_from_this());
+            } else {
+                depth--;
+                parent.zk.create(path_parent(path, depth), "", false, false, shared_from_this());
+            }
+        } else if(reply.rc == ZNONODE) {
+            depth++;
+            parent.zk.create(path_parent(path, depth), "", false, false, shared_from_this());
+        } else {
+            throw error_t(map_zoo_error(reply.rc), "failure during creating node - {}", zerror(reply.rc));
+        }
+    }
+
+    auto on_reply(children_reply_t reply) -> void override {
+        if(reply.rc) {
+            throw error_t(map_zoo_error(reply.rc), "failed to get lock folder children - {}", zerror(reply.rc));
+        }
+        auto& children = reply.children;
+        std::sort(children.begin(), children.end());
+        COCAINE_LOG_DEBUG(parent.log, "children - {}, created path - {}", children, created_sequence_node);
+        auto it = std::find(children.begin(), children.end(), created_sequence_node);
```

Comment:
Not checked, but `lower_bound` _could_ perform a bit better on large vectors (seems not a case at all here).